base/set.jl

```diff
@@ -427,6 +427,34 @@ allunique(::Union{AbstractSet,AbstractDict}) = true
 
 allunique(r::AbstractRange) = !iszero(step(r)) || length(r) <= 1
 
+function allunique(A::AbstractArray)
+    if length(A) < 2
+        return true
+    elseif length(A) < 25 && IndexStyle(A) isa IndexLinear
+        # then linear search is faster, even in worst case
```

Comment:
On my local bench with `Vector{Float64}`, the following seems faster.
```julia
iter = eachindex(A)
I = iterate(iter)
while I !== nothing
    i, s = I
    a = A[i]
    for j in Iterators.rest(iter, s)
        isequal(a, A[j]) && return false
    end
    I = iterate(iter, s)
end
return true
```

Reply:
Thanks for looking, will check some more.

I thought I convinced myself that an *unconditional* inner loop usually was faster than one with a return option, but could be wrong; `any` also has logic for `missing` which isn't needed. I made the loop do the short runs first, so that the case of all equal is fast.

Is `Iterators.rest` still be fast without `IndexLinear`, e.g. for transposed arrays?

Reply:
I tested with `randn(12,2)'`, current `Base` cost 435ns and the above cost 440ns. (quite close.)
The main problem might be `mappedarray` with heavy `f`, i.e. `sin`.
We can't always assume `getindex` is cheap.

Reply:
All of these approaches are potentially awful for `mappedarray`, right? Iterate won't cache the values. It claims to support fast linear indexing, so that's not going to save us... is there some other trait for this?
```
julia> mappedarray(log, 1:3) |> Base.IndexStyle
IndexLinear()
```
We could restrict to StridedArray or something.

Reply:
I've updated the PR to use this now. Many thanks, it's quite a bit faster than what I had. Times:
```julia
julia> for n in 10:10:50
         @show n
         x = rand(n) 
         @btime allunique($x)
         @btime alluniquePR($x) 
         @btime alluniqueNN($x)
       end
n = 10
  min 110.247 ns, mean 126.389 ns (4 allocations, 400 bytes)
  min 47.354 ns, mean 47.943 ns (0 allocations)
  min 32.696 ns, mean 33.019 ns (0 allocations)
n = 20
  min 258.878 ns, mean 298.428 ns (7 allocations, 1.12 KiB)
  min 175.492 ns, mean 176.568 ns (0 allocations)
  min 125.743 ns, mean 126.865 ns (0 allocations)
n = 30
  min 315.988 ns, mean 358.934 ns (7 allocations, 1.12 KiB)
  min 386.901 ns, mean 390.457 ns (0 allocations)
  min 281.505 ns, mean 283.823 ns (0 allocations)
n = 40
  min 393.772 ns, mean 440.944 ns (7 allocations, 1.12 KiB)
  min 693.878 ns, mean 701.657 ns (0 allocations)
  min 507.984 ns, mean 513.435 ns (0 allocations)
n = 50
  min 669.057 ns, mean 1.022 μs (10 allocations, 3.62 KiB)
  min 1.067 μs, mean 1.085 μs (0 allocations)
  min 801.966 ns, mean 809.811 ns (0 allocations)

julia> for n in 10:10:50
         @show n
         x = rand(1:3, n) 
         @btime allunique($x)
         @btime alluniquePR($x) 
         @btime alluniqueNN($x)
       end
n = 10
  min 72.328 ns, mean 88.557 ns (4 allocations, 400 bytes)
  min 6.791 ns, mean 6.901 ns (0 allocations)
  min 3.958 ns, mean 4.067 ns (0 allocations)
n = 20
  min 75.275 ns, mean 91.740 ns (4 allocations, 400 bytes)
  min 9.718 ns, mean 9.834 ns (0 allocations)
  min 5.208 ns, mean 5.317 ns (0 allocations)
n = 30
  min 72.382 ns, mean 89.386 ns (4 allocations, 400 bytes)
  min 6.791 ns, mean 6.883 ns (0 allocations)
  min 3.666 ns, mean 3.758 ns (0 allocations)
n = 40
  min 75.317 ns, mean 92.798 ns (4 allocations, 400 bytes)
  min 9.718 ns, mean 9.853 ns (0 allocations)
  min 3.625 ns, mean 3.756 ns (0 allocations)
n = 50
  min 72.296 ns, mean 89.128 ns (4 allocations, 400 bytes)
  min 6.791 ns, mean 6.916 ns (0 allocations)
  min 3.333 ns, mean 3.456 ns (0 allocations)

julia> for n in 5:5:25
         @show 2n
         x = rand(2, n)' 
         @btime allunique($x)
         @btime alluniquePR($x) 
         @btime alluniqueNN($x)
       end
2n = 10
  min 111.892 ns, mean 129.408 ns (4 allocations, 400 bytes)
  min 169.850 ns, mean 170.534 ns (0 allocations)
  min 63.583 ns, mean 63.838 ns (0 allocations)
2n = 20
  min 267.415 ns, mean 304.367 ns (7 allocations, 1.12 KiB)
  min 538.577 ns, mean 541.051 ns (0 allocations)
  min 197.484 ns, mean 198.860 ns (0 allocations)
2n = 30
  min 316.847 ns, mean 355.317 ns (7 allocations, 1.12 KiB)
  min 1.087 μs, mean 1.104 μs (0 allocations)
  min 423.367 ns, mean 424.834 ns (0 allocations)
2n = 40
  min 396.243 ns, mean 437.514 ns (7 allocations, 1.12 KiB)
  min 1.775 μs, mean 1.790 μs (0 allocations)
  min 749.320 ns, mean 761.088 ns (0 allocations)
2n = 50
  min 673.745 ns, mean 1.026 μs (10 allocations, 3.62 KiB)
  min 2.718 μs, mean 2.760 μs (0 allocations)
  min 1.163 μs, mean 1.190 μs (0 allocations)
```